File: TsinghuaNet.Helper/Models/NetCommand.vb
Imports System.ComponentModel
Imports System.Windows.Input

Public Class NetCommand
    Implements ICommand

    Private ReadOnly model As NetViewModel
    Private executor As Func(Of IConnect, Task(Of LogResponse))

    Public Sub New(model As NetViewModel, executor As Func(Of IConnect, Task(Of LogResponse)))
        Me.model = model
        Me.executor = executor
        AddHandler model.PropertyChanged, AddressOf OnModelPropertyChanged
        AddHandler model.Credential.PropertyChanged, AddressOf OnModelPropertyChanged
    End Sub

    Private Sub OnModelPropertyChanged(sender As Object, e As PropertyChangedEventArgs)
        If e.PropertyName = "State" OrElse e.PropertyName = "IsBusy" Then
            OnCanExecuteChanged()
        End If
    End Sub

    Public Event CanExecuteChanged As EventHandler Implements ICommand.CanExecuteChanged
    Protected Overridable Sub OnCanExecuteChanged()
        RaiseEvent CanExecuteChanged(Me, EventArgs.Empty)
    End Sub

    Public Function CanExecute(parameter As Object) As Boolean Implements ICommand.CanExecute
        Return Not model.IsBusy AndAlso model.Credential.State <> NetState.Unknown
    End Function

    Public Async Sub Execute(parameter As Object) Implements ICommand.Execute
        Await model.NetCommandExecuteAsync(executor)
    End Sub
End Class

File: TsinghuaNet/Models/NetCommand.cs
using System;
using System.ComponentModel;
using System.Threading.Tasks;
using System.Windows.Input;
using TsinghuaNet.Helpers;
using TsinghuaNet.ViewModels;

namespace TsinghuaNet.Models
{
    public class NetCommand : ICommand
    {
        private readonly MainViewModelBase model;
        private Func<IConnect, Task<LogResponse>> executor;

        public NetCommand(MainViewModelBase model, Func<IConnect, Task<LogResponse>> executor)
        {
            this.model = model;
            this.executor = executor;
            model.PropertyChanged += OnModelPropertyChanged;
            model.Credential.PropertyChanged += OnModelPropertyChanged;
        }

        private void OnModelPropertyChanged(object sender, PropertyChangedEventArgs e)
        {
            if (e.PropertyName == "State" || e.PropertyName == "IsBusy")
                OnCanExecuteChanged();
        }

        public event EventHandler CanExecuteChanged;
        protected virtual void OnCanExecuteChanged() => CanExecuteChanged?.Invoke(this, EventArgs.Empty);

        public bool CanExecute(object parameter) => !model.IsBusy && model.Credential.State != NetState.Unknown;

        public async void Execute(object parameter) => await model.NetCommandExecuteAsync(executor);
    }
}
